### ISSUE TITLE
Rust-Postgres, binary support - ANTLR, Can you explain the parsing workflow?

### ISSUE THREAD

**Issue by afidegnum** [apache/incubator-age#162]
I'm struggling to understand the concept of parsing with ANTLR and mapping the AGE output to custom data type. 

Can you please throw more light?

My Initial attempt of executing the query brought the earlier error of: `Error: db error: ERROR: no binary output function available for type agtype!`

I would need to cast `agtype` to a readable format.

**Comment by afidegnum** [comment]
any update?

**Comment by emotionbug** [comment]
I'll look into your issue with #158

**Comment by emotionbug** [comment]
I found some thread about this.

when use rust-postgres driver needs binary I/O.
I think It needs some implementation.

@jrgemignani @JoshInnis 

https://github.com/sfackler/rust-postgres/issues/389
https://www.postgresql.org/message-id/CANmj9Vz10xO_%2BUTE_K7o7BcdDpZhc0W6xBYB%2B7%2Bz6%2BfLVdoxMQ%40mail.gmail.com

**Comment by afidegnum** [comment]
This is my attempted implementation so far, which wasn't successful. I believe there is something I'm not properly addressing. 

Cargo.toml
```
 . . . . 
# [dependencies]
rand = { version = "0.7.3" }
tokio-postgres = { version = "0.5", features = ["with-chrono-0_4", "with-uuid-0_8"] }
postgres-protocol = "*"
postgres = "0.19.2"
postgres-types = "*"

```
src/main.rs

```
struct AgType<'a>(&'a str);

impl<'a> Display for AgType<'a>{
    
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            let d: &str = self.0 ; 
            write!(f, "({})", d)
    }
}

impl<'a> FromSql<'a> for AgType<'a> {
    fn from_sql(_: &Type, raw: &'a [u8]) -> Result<AgType<'a>, Box<dyn Error + Sync + Send>> {
        let t = types::text_from_sql(raw)?;
        Ok(AgType(t))
    }

    fn accepts(ty: &Type) -> bool {
        ty.name() == "agtype"
    }
}


fn main(){

pub fn graph_list() -> Result<(), Box<dyn Error>> {
    let conn_string = "host=localhost port=5432 dbname=texttest user=afidegnum password=chou1979";
    let mut client= Client::connect(conn_string, NoTls)?;
    
    &client
        .batch_execute("LOAD \'age\'")?;

    &client
        .batch_execute("SET search_path = ag_catalog, \"$user\", public")?;

     for row in &client.query("SELECT count(*) FROM ag_graph WHERE name= 'text_test'", &[])?{
         println!("{}", row.get::<_, i64>(0));
         let  count: i64 = row.get(0); println!("{}", count);
         println!("{}", count);
         dbg!(count);
     }

    

    for row in client.query("SELECT * from cypher('text_test', $$ MATCH (V:node) RETURN id(V) as vid $$) as (vid agtype)", &[])? {
        let path: AgType = row.get(0);
        println!("{}", path)
     }
    Ok(())
}


let res = graph_list();
match res {
        Ok(r) => {println!("Succeeded! {:?} ", r);},
        Err(e) => {println!("Error: {}!", e);}
    }



}
```
```
1
[src/lib.rs:20] count = 1
1
1
Error: db error: ERROR: no binary output function available for type agtype!
```

**Comment by emotionbug** [comment]
prepare
```pgsql
SELECT * from create_graph('text_test')
SELECT * from cypher('text_test', $$ CREATE (V:node{id:1}) $$) as (V agtype
```

If you use that commit to execute the above code, output is as follows.
output 
```
Row { columns: [Column { name: "count", type: Int8 }] }
"{\"id\": 844424930131969, \"label\": \"node\", \"properties\": {\"id\": 1}}::vertex"
Succeeded! () 
```

**Comment by afidegnum** [comment]
using rust-postgres or from the console ?

**Comment by emotionbug** [comment]
With rust-postgres

**Comment by afidegnum** [comment]
ok, give me 15mins, let me check right away

**Comment by afidegnum** [comment]
The issue i'm having is querying the result, that's where i'm having the binary error

**Comment by emotionbug** [comment]
@afidegnum right.
If you can, checkout that PR. and, drop extension, reinstall extension.

because, only reinstall will not affect `pg_type` definition.

**Comment by afidegnum** [comment]
ok, good, right away!

**Comment by afidegnum** [comment]
what's the command for checking out the binary branch ?

**Comment by emotionbug** [comment]
`Git branch` show current branch.

**Comment by afidegnum** [comment]
master

**Comment by emotionbug** [comment]
Change branch. `git checkout binary_io`

**Comment by afidegnum** [comment]
ok, wait a min

**Comment by afidegnum** [comment]
still no luck, even though used the `binary_io` branch I even restarted the PostgreSQL server

**Comment by afidegnum** [comment]
the error lies when i query results like 
```sql
SELECT * from cypher('text_test', $$
          MATCH (V:node)
          RETURN V
$$) as (V agtype);
```

**Comment by afidegnum** [comment]
can we capitalize on something like this? 

`COPY foo FROM stdin BINARY`

**Comment by afidegnum** [comment]
any insight ?

**Comment by emotionbug** [comment]
I tested again. it works.

```sh
git clone -b binary_io --single-branch https://github.com/emotionbug/incubator-age.git
make -j16 && make install -j16
```

```psql
update pg_type set typsend = 'ag_catalog.agtype_send', typreceive = 'ag_catalog.agtype_recv' where typname = 'agtype';
```

```rust
use postgres::{Client, Error, NoTls};
use postgres::types::{FromSql, Type};

struct AgType<'a>(&'a str);


impl<'a> FromSql<'a> for AgType<'a> {
    fn from_sql(_: &Type, raw: &'a [u8]) -> Result<AgType<'a>, Box<dyn std::error::Error + Sync + Send>> {
        let t = postgres_protocol::types::text_from_sql(raw)?;
        Ok(AgType(t))
    }

    fn accepts(ty: &Type) -> bool {
        ty.name() == "agtype"
    }
}

pub fn graph_list() -> Result<(), Error> {
    let conn_string = "host=172.17.0.2 port=5432 dbname=postgres user=postgres password=postgres";
    let mut client = Client::connect(conn_string, NoTls)?;


    &client.batch_execute("LOAD \'age\'")?;
    &client.batch_execute("SET search_path = ag_catalog, \"$user\", public")?;
    // &client.batch_execute("SELECT * from create_graph('text_test')")?;
    // &client.batch_execute("SELECT * from cypher('text_test', $$ CREATE (V:node{id:1}) $$) as (V agtype)")?;

    for types in &client.query("SELECT count(*) FROM ag_graph WHERE name= 'text_test'", &[])? {
        println!("{:?}", types)
    }


    for row in client.query("SELECT * from cypher('text_test', $$ MATCH (V:node) RETURN V $$) as (V agtype)", &[])? {
        let path: AgType = row.get(0);
        println!("{:?}", path.0);
    }
    Ok(())
}

fn main() {
    match graph_list() {
        Ok(r) => { println!("Succeeded! {:?} ", r); }
        Err(e) => { println!("Error: {}!", e); }
    }
}

```

**Comment by afidegnum** [comment]
Thanks, let me try again

**Comment by afidegnum** [comment]
it's a bit strange, I'm still not lucky even though the query returned a successful result. 

I could see you run your extension from the docker. I'm using Arch Linux, is there something that needs implementing? 

```sql
postgres=# select typreceive, typsend from pg_type where typname = 'agtype';
       typreceive       |        typsend         
------------------------+------------------------
 ag_catalog.agtype_recv | ag_catalog.agtype_send
(1 row)

```

```rs
Row { columns: [Column { name: "count", type: Int8 }] }
Error: db error: ERROR: no binary output function available for type agtype!
```

**Comment by emotionbug** [comment]
@afidegnum


```Dockerfile
FROM postgres:11

RUN apt-get update 
RUN apt-get install --assume-yes --no-install-recommends --no-install-suggests \
  bison \
  build-essential \
  ca-certificates \
  flex \
  git \
  postgresql-plpython3-11 \
  postgresql-server-dev-11 

RUN git clone -b binary_io --single-branch https://github.com/emotionbug/incubator-age.git /age 
RUN cd /age && make install 
```

I always use docker environment. and, it works as above scenario without any problems in a new environment.
I think you are missed something.

**Comment by afidegnum** [comment]
what about the permissions? can all be done at the root level?

**Comment by emotionbug** [comment]
@afidegnum I don't know why are you talking about permission. but, it can

**Comment by afidegnum** [comment]
have you tried fetching results ? i.e fetch nodes

**Comment by afidegnum** [comment]
the problem i'm having is when running this particular query 

```rs
 for row in client.query("SELECT * from cypher('text_test', $$ MATCH (V:node) RETURN V $$) as (V agtype)", &[])? {
        let path: AgType = row.get(0);
        println!("{:?}", path.0);
    }
    
```

**Comment by emotionbug** [comment]
@afidegnum  check on above comment. I wrote my code, and results.

**Comment by jrgemignani** [comment]
I have verified that the message is output from PG functions and will look into the possibility of add it. Unfortunately, I cannot make any guarantees due to the nature of PGs code base.

**Comment by afidegnum** [comment]
Thanks a lot

**Comment by jrgemignani** [comment]
I'm going to have to retract that. After rereading the thread, everything that @emotionbug has done is correct. My response was to an earlier comment so, it doesn't apply here.

**Comment by jrgemignani** [comment]
I have pushed to the master branch @emotionbug and my patch. It "should" correct the issue for anyone using the latest clone of the master. Please update AGE on your side after cloning and building from the master to try it out.

**Comment by afidegnum** [comment]
ok, Everything is working when I used @emotionbug repo in docker, I will need to try the fresh installation on my Arch distro too. that's where i have been having the problem.

**Comment by emotionbug** [comment]
I close this issue because this can be reference in https://github.com/apache/incubator-age/issues/207
